module/zstd/zstd.c

```diff
@@ -208,37 +203,37 @@ zstd_mempool_alloc(struct zstd_pool *zstd_mempool, size_t size)
 			mutex_exit(&pool->barrier);
 		}
 	}
+
+	if (mem)
+		return (mem);
 	/*
 	 * if no preallocated slot was found, try to fill in a new one
 	 */
-	if (!mem) {
-		for (i = 0; i < ZSTD_POOL_MAX; i++) {
-			pool = &zstd_mempool[i];
-			if (mutex_tryenter(&pool->barrier)) {
-				/* object is free, try to allocate new one */
-				if (!pool->mem) {
-					mem = vmem_alloc(size, KM_SLEEP);
-					pool->mem = mem;
-					/* allocation successfull? */
-					if (pool->mem) {
-						/*
-						 * keep track for later release
-						 * and update timestamp
-						 */
-						mem->pool = pool;
-						pool->size = size;
-						pool->timeout =
-						    gethrestime_sec() +
-						    ZSTD_POOL_TIMEOUT;
-						mem->kmem_type = ZSTD_KMEM_POOL;
-						mem->kmem_size = size;
-					} else {
-						mutex_exit(&pool->barrier);
-					}
-					return (mem);
+	for (i = 0; i < ZSTD_POOL_MAX; i++) {
+		pool = &zstd_mempool[i];
+		if (mutex_tryenter(&pool->barrier)) {
+			/* object is free, try to allocate new one */
+			if (!pool->mem) {
+				mem = vmem_alloc(size, KM_SLEEP);
+				pool->mem = mem;
+				/* allocation successfull? */
+				if (pool->mem) {
+					/*
+					 * keep track for later release
+					 * and update timestamp
+					 */
+					mem->pool = pool;
+					pool->size = size;
+					pool->timeout = gethrestime_sec() +
+					    ZSTD_POOL_TIMEOUT;
+					mem->kmem_type = ZSTD_KMEM_POOL;
+					mem->kmem_size = size;
+				} else {
+					mutex_exit(&pool->barrier);
 				}
-				mutex_exit(&pool->barrier);
+				return (mem);
 			}
+			mutex_exit(&pool->barrier);
 		}
 	}
 
```

Comment:
I wonder if we can ever get here with anything else than !mem? when allocation was successfull, we already return in line 234.

Reply:
yes you get there if allocation fails. so what? i mean if the object cannot be allocated, it will be released for reusing or next try. we cannot keep it locked since it will be lost forever then

Reply:
so a object which is unused must be released. if its taken, it will be locked and released later by zstd_free

Reply:
of course we can get some speedup here if KM_SLEEP would not be a deadlock operation like i suggested using the KM_ONCE flag. (which is not part of the source anymore right now) but i will get back to this topic once its merged. so if GFP_KERNEL / KM_SLEEP allocation fails it will not try forever to get the memory, but exit with NULL. so the mutex_exit will trigger then

Reply:
please read again what I asked. you have a complete ... what is "Themaverfehlung" in english...?

Reply:
if allocation was successfull you never get there if mem != NULL. but whats the question? i mean you should not get there and this is wanted

Reply:
or you you refer to !pool->mem ?

Reply:
the question is right. but consider the case that the first free mutex is a preallocated object. so its non null and need to be ignored. since we checked for preallocated free objects in the first for loop. now on the second try we did not found any and try to allocate a new one in the first slot. but between both loops zstd_free could be already called from another thread, so we get a race. so we need to check if pool is NULL and if its not, we just ignore the object get to the next slot

Reply:
in case of a case we can also do the following
`--- a/module/zstd/zstd.c
+++ b/module/zstd/zstd.c
@@ -233,7 +233,7 @@ zstd_mempool_alloc(struct zstd_pool *zstd_mempool, size_t size)
                                }
                                return (mem);
                        }
-                       mutex_exit(&pool->barrier);
+                       return pool->mem;
                }
        }

`

Reply:
mutex_exit can be replaced by return (pool->mem); thats one possible option. just in case there is a new free one

Reply:
line 240!!

Reply:
~~~
blah

240!!!!!! if (!mem) {
  blah
  }

return (mem);
~~~

Reply:
ping